File: metal/paddle-mobile/paddle-mobile/Loader.swift
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License. */

import Foundation
import SwiftProtobuf


public class Loader<P: PrecisionType> {
    class ParaLoader {
        let file: UnsafeMutablePointer<FILE>
        let fileSize: Int
        var nowIndex: Int
        init(paramPath: String) throws {
            guard let tmpFile = fopen(paramPath, "rb") else {
                throw PaddleMobileError.loaderError(message: "open param file error" + paramPath)
            }
            file = tmpFile
            fseek(file, 0, SEEK_END)
            fileSize = ftell(file)
            guard fileSize > 0 else {
                throw PaddleMobileError.loaderError(message: "param file size is too small")
            }
            rewind(file)
            nowIndex = 0
        }
        
        func read(tensor: Tensor<P>) throws {
            guard nowIndex <= fileSize else {
                throw PaddleMobileError.loaderError(message: "out of the file range")
            }
            
            func pointerReader<T>(type: T.Type) -> T {
                let ptr = UnsafeMutablePointer<T>.allocate(capacity: MemoryLayout<T>.size)
                fread(ptr, 1, MemoryLayout<T>.size, file)
                nowIndex += MemoryLayout<T>.size
                let pointee = ptr.pointee
                ptr.deinitialize(count: MemoryLayout<UInt32>.size)
                ptr.deallocate()
                return pointee
            }
            
            _ = pointerReader(type: UInt32.self)
            let lodLevel = pointerReader(type: UInt64.self)
            for _ in 0..<lodLevel {
                let size = pointerReader(type: UInt64.self)
                for _ in 0..<Int(size/UInt64(MemoryLayout<size_t>.size)){
                    _ = pointerReader(type: size_t.self)
                }
            }
            
            let _ = pointerReader(type: UInt32.self)
            
            let tensorDescSize = pointerReader(type: Int32.self)
            fseek(file, Int(tensorDescSize), SEEK_CUR)
            nowIndex += Int(tensorDescSize)
            
            /*
             这里没有根据 Data Type 去判断, 而是从外部泛型直接指定了精度
             */

            let bytesRead = fread(tensor.data.pointer, 1, tensor.data.size, file)
            guard bytesRead == tensor.data.size else {
                throw PaddleMobileError.loaderError(message: "param read size error")
            }
            nowIndex += bytesRead
        }
        
        deinit {
            fclose(file)
        }
    }
    public init(){}
    public func load(modelPath: String, paraPath: String) throws -> Program{
        guard let modelData = try? Data.init(contentsOf: URL.init(fileURLWithPath: modelPath)) else {
            throw PaddleMobileError.loaderError(message: "load " + modelPath + " failed !")
        }
        
        do {
            let protoProgram = try PaddleMobile_Framework_Proto_ProgramDesc.init(
                serializedData: modelData)
            let scope = Scope.init()
            let programDesc = ProgramDesc.init(protoProgram: protoProgram)
            
            guard let paraLoader = try? ParaLoader.init(paramPath: paraPath) else {
                throw PaddleMobileError.loaderError(message: "load para error")
            }
            
            guard programDesc.blocks.count > 0 else {
                throw PaddleMobileError.loaderError(message: "count of blocks must greater than 0")
            }
            
            // to get feed key and fetch key
            let block = programDesc.blocks[0]
            guard let firstOp = block.ops.first, let lastOp = block.ops.last else {
                throw PaddleMobileError.loaderError(message: "at least two operator")
            }
            guard firstOp.type == gFeedType, lastOp.type == gFetchType else {
                throw PaddleMobileError.loaderError(message: "the first op is not feed or the last op is not fetch")
            }
            
            guard let inputKey = opInfos[gFeedType]?.inputs.first, let outKey = opInfos[gFetchType]?.outputs.first else {
                throw PaddleMobileError.loaderError(message: "the feed input key or fetch output key not found")
            }
            guard let feedKey = firstOp.inputs[inputKey]?.first, let fetchKey = lastOp.outputs[outKey]?.first else {
                throw PaddleMobileError.loaderError(message: "feed key or fetch key not found")
            }
            
            // to load memory
            for block in programDesc.blocks {
                for varDesc in block.vars {
                    if (varDesc.type == .LodTensor) {
                        guard let tensorDesc = varDesc.tensorDesc else {
                            throw PaddleMobileError.loaderError(message: "get tensor desc failed")
                        }
                        
                        guard (try? tensorDesc.dataType.dataTypeSize()) == MemoryLayout<P>.size else {
                            throw PaddleMobileError.memoryError(message: "PrecisionType not support")
                        }
                        
                        if (varDesc.persistable
                            && varDesc.type != .FeedMiniBatch
                            && varDesc.type != .FetchList) {
                            let dimArr = tensorDesc.dims
                            
                            guard dimArr.count > 0 else {
                                throw PaddleMobileError.loaderError(message: "tensor desc dim size error")
                            }
                            
                            let dim = Dim.init(inDim: dimArr)
                            let tensor = Tensor<P>.init(inDim: dim, inLayout: tensorDesc.dataLayout)
                            do {
                                try paraLoader.read(tensor: tensor)
                            } catch let error {
                                throw error
                            }
                            tensor.convert(to: .NHWC)
                            scope[varDesc.name] = tensor
                        } else {
                            let dim = Dim.init(inDim: tensorDesc.NHWCDim)
                            scope[varDesc.name] = Texture.init(inDim: dim, inLayout: .NHWC)
                        }
                    } else {
                        if varDesc.name == fetchKey {
                            scope[varDesc.name] = ResultHolder<P>.init(inDim: [], inResult: [])
                        } else if varDesc.name == feedKey {
                            scope[varDesc.name] = Texture.init()
                        }
                    }
                }
            }
            
            let program = Program.init(protoProgramDesc: protoProgram, inParamPath: paraPath, inScope: scope, inFeedKey: feedKey, inFetchKey: fetchKey)
            
            return program
        } catch _ {
            throw PaddleMobileError.loaderError(message: "protobuf decoder error")
        }
    }
}

File: src/io/loader.cpp
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "io/loader.h"
#include <model-decrypt/include/model_crypt.h>
#include <cstring>

#include "framework/lod_tensor.h"
#include "framework/program/program-optimize/program_optimize.h"
#define ENABLE_CRYPT
#ifdef ENABLE_CRYPT

#include "model-decrypt/include/model_decrypt.h"


char encrypted_key[] = {0xfd, 0xa9, 0xfa, 0xbe, 0xf3, 0xed, 0xe0, 0x87, 0xa7,
                        0xa4, 0xab, 0xbb, 0xeb, 0xb8, 0xfb, 0xed, 0xbf, 0xed,
                        0xa4, 0x87, 0x9e, 0x81, 0xf7, 0xa9, 0x87, 0x84, 0xef,
                        0xf8, 0xbd, 0xeb, 0x90, 0xb2, 0xb6, 0xfd, 0xb2, 0x8b,
                        0xb3, 0xa7, 0xf5, 0x95, 0xf5, 0xe5, 0x91, 0xe9, 0xab,
                        0xaf, 0x93, 0x8c, 0xac, 0xbf};

static size_t ReadBufferCrypt(const char *file_path, uint8_t **decrypt_output) {
  std::string str = file_path;
  std::cout << "in!!!!" << str << std::endl;
  int ret = 0;
  void *context = NULL;
  unsigned int sign = 0;
  void *file_map = NULL;
  unsigned int file_size = 0;
  //    unsigned char *decrypt_output = *out;
  unsigned int decrypt_output_size = 0;
  DLOG << "sizeof(encrypted_key): " << sizeof(encrypted_key);
  for (int i = 0; i < sizeof(encrypted_key); ++i) {
    //  DLOG<<*(encrypted_key+i);
    encrypted_key[i] ^= 0xde;
    DLOG << *(encrypted_key + i);
  }

  // init decryption context
  ret = init_crypt_context((const unsigned char *)encrypted_key,
                           static_cast<unsigned int>(strlen(encrypted_key)),
                           &context, &sign);
  if (0 != ret) {
    std::cout << "failed to init_crypt_context." << std::endl;
    return static_cast<size_t>(-1);
  }
  std::cout << "init_crypt_context succeed.";

  // create file mapping for encrypted model file
  ret = open_file_map(file_path, &file_map, &file_size);
  if (0 != ret) {
    char buf[128];
    snprintf(buf, sizeof(buf), "failed to open_file_map for file %s",
             file_path);
    std::cout << "failed to open_file_map.\n";
    return static_cast<size_t>(-1);
  }
  std::cout << "open_file_map succeed.  file_size: " << file_size << std::endl;

  // decrypt
  ret = model_decrypt(context, sign, (unsigned char *)file_map, file_size,
                      decrypt_output, &decrypt_output_size);
  if (0 != ret) {
    printf("failed to model_decrypt.\n");
    std::cout << "failed to model_decrypt.  ret:" << ret << std::endl;
    return static_cast<size_t>(-1);
  }
  std::cout << "model_decrypt succeed." << std::endl;
  std::cout << "decrypt_output_size: " << decrypt_output_size << std::endl;

  // save decrypted file to /sdcard
  //        ret = save_buf_to_file(decrypt_output, decrypt_output_size,
  //        decrypted_file_path); if (0 != ret) {
  //            printf("failed to save_buf_to_file.\n");
  //            result.no = -1;
  //            result.description = "failed to save_buf_to_file.";
  //            return result;
  //        }
  //        DLOG<<"save_buf_to_file succeed.";

  // release output
  //    free(decrypt_output);
  //    decrypt_output = NULL;

  // release file mapping
  close_file_map(file_map, file_size);
  file_map = NULL;

  // release decryption context
  uninit_crypt_context(context);
  context = NULL;

  //    result.no = 0;
  //    result.description = "decrypt successfully!";
  return decrypt_output_size;
}

#endif

namespace paddle_mobile {
using framework::Variable;

static size_t ReadBuffer(const char *file_name, uint8_t **out) {
  FILE *fp;
  fp = fopen(file_name, "rb");
  PADDLE_MOBILE_ENFORCE(fp != NULL, " %s open failed !", file_name);

  fseek(fp, 0, SEEK_END);
  size_t size = ftell(fp);
  rewind(fp);

  DLOG << "model size: " << size;

  *out = reinterpret_cast<uint8_t *>(malloc(size));

  size_t cur_len = 0;
  size_t nread;
  while ((nread = fread(*out + cur_len, 1, size - cur_len, fp)) != 0) {
    cur_len += nread;
  }
  fclose(fp);
  return cur_len;
}

template <typename Dtype, Precision P>
const framework::Program<Dtype, P> Loader<Dtype, P>::Load(
    const std::string &dirname, bool optimize, bool quantification,
    bool can_add_split) {
#ifdef ENABLE_CRYPT
  auto program = this->LoadProgram(dirname + "/model.mlm", optimize,
                                   quantification, can_add_split);
  program.model_path = dirname;
  return program;
#else
  auto program = this->LoadProgram(dirname + "/__model__", optimize,
                                   quantification, can_add_split);
  program.model_path = dirname;
  return program;
#endif
}

template <typename Dtype, Precision P>
const framework::Program<Dtype, P> Loader<Dtype, P>::Load(
    const std::string &model_path, const std::string &para_path, bool optimize,
    bool quantification) {
  auto program = this->LoadProgram(model_path, optimize, quantification);

  program.para_path = para_path;
  program.combined = true;
  program.quantification = quantification;
  return program;
}

template <typename Dtype, Precision P>
const framework::Program<Dtype, P> Loader<Dtype, P>::LoadProgram(
    const std::string &model_path, bool optimize, bool quantification,
    bool can_add_split) {
  std::string model_filename = model_path;
  PaddleMobile__Framework__Proto__ProgramDesc *c_program;
  uint8_t *buf = NULL;

  size_t read_size;
#ifdef ENABLE_CRYPT
  read_size = ReadBufferCrypt(model_filename.c_str(), &buf);
#else
  read_size = ReadBuffer(model_filename.c_str(), &buf);
#endif

  PADDLE_MOBILE_ENFORCE(buf != NULL, "read from __model__ is null");

  for (int i = 0; i < read_size; ++i) {
    DLOG << "buf -" << i << *(buf + i);
  }

  c_program = paddle_mobile__framework__proto__program_desc__unpack(
      NULL, read_size, buf);

  DLOG << "read_size: " << read_size;
  DLOG << "buf: " << *buf;
  //
  PADDLE_MOBILE_ENFORCE(c_program != NULL, "program is null");
  //
  DLOG << "n_ops: " << (*c_program->blocks)->n_ops;
  //
  auto originProgramDesc = std::make_shared<framework::ProgramDesc>(c_program);

  framework::Program<Dtype, P> program;
  program.originProgram = originProgramDesc;
  program.quantification = quantification;

  auto scope = std::make_shared<framework::Scope>();
  program.scope = scope;

  for (const auto &block : originProgramDesc->Blocks()) {
    for (auto var_desc : block->Vars()) {
      auto var = scope->Var(var_desc->Name());

      if (var_desc->Type() == framework::VARTYPE_TYPE_LOD_TENSOR) {
        if (var_desc->Persistable() &&
            var_desc->Type() != framework::VARTYPE_TYPE_FEED_MINIBATCH &&
            var_desc->Type() != framework::VARTYPE_TYPE_FETCH_LIST) {
          auto dim = var_desc->Tensor_desc().Dims();
          auto tensor = var->GetMutable<framework::LoDTensor>();
          tensor->Resize(framework::make_ddim(dim));
        } else {
          auto dim = var_desc->Tensor_desc().Dims();
          PADDLE_MOBILE_ENFORCE(dim.size() > 0, "dim size is 0");
          dim[0] = 1;
          auto tensor = var->GetMutable<framework::LoDTensor>();
          tensor->Resize(framework::make_ddim(dim));
        }
      } else {
        // TODO(codeWorm): some.
      }
    }
  }

  if (optimize) {
    framework::ProgramOptimize program_optimize;
    program.optimizeProgram =
        program_optimize.FusionOptimize(originProgramDesc, can_add_split);
  }
  if (optimize) {
    program.optimizeProgram->Description("optimize: ");
  } else {
    originProgramDesc->Description("program: ");
  }

  paddle_mobile__framework__proto__program_desc__free_unpacked(c_program, NULL);
  return program;
}

template class Loader<CPU, Precision::FP32>;
template class Loader<FPGA, Precision::FP32>;
template class Loader<GPU_MALI, Precision::FP32>;

}  // namespace paddle_mobile
